lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -116,12 +117,311 @@ void printCountAsAll(OpAsmPrinter &printer, Operation *op, IntegerAttr count) {
 // Substrait operations
 //===----------------------------------------------------------------------===//
 
+namespace mlir {
+namespace substrait {
+
+static ParseResult
+parseAggregationInvocation(OpAsmParser &parser,
+                           AggregationInvocationAttr &aggregationInvocation);
+static void
+printAggregationInvocation(OpAsmPrinter &printer, CallOp op,
+                           AggregationInvocationAttr aggregationInvocation);
+static ParseResult parseAggregateRegions(OpAsmParser &parser,
+                                         Region &groupingsRegion,
+                                         Region &measuresRegion,
+                                         ArrayAttr &groupingSetsAttr);
+static void printAggregateRegions(OpAsmPrinter &printer, AggregateOp op,
+                                  Region &groupingsRegion,
+                                  Region &measuresRegion,
+                                  ArrayAttr groupingSetsAttr);
+
+} // namespace substrait
+} // namespace mlir
+
 #define GET_OP_CLASSES
 #include "substrait-mlir/Dialect/Substrait/IR/SubstraitOps.cpp.inc"
 
 namespace mlir {
 namespace substrait {
 
+ParseResult
+parseAggregationInvocation(OpAsmParser &parser,
+                           AggregationInvocationAttr &aggregationInvocation) {
+  // This is essentially copied from `FieldParser<AggregationInvocation>` but
+  // sets the default `all` case if no invocation type is present.
+
+  MLIRContext *context = parser.getContext();
+  std::string keyword;
+  if (failed(parser.parseOptionalKeywordOrString(&keyword))) {
+    // No keyword parse --> use default value.
+    aggregationInvocation =
+        AggregationInvocationAttr::get(context, AggregationInvocation::all);
+    return success();
+  }
+
+  // Symbolize the keyword.
+  if (std::optional<AggregationInvocation> attr =
+          symbolizeAggregationInvocation(keyword)) {
+    aggregationInvocation =
+        AggregationInvocationAttr::get(parser.getContext(), attr.value());
+    return success();
+  }
+
+  // Symbolization failed.
+  auto loc = parser.getCurrentLocation();
+  return parser.emitError(loc)
+         << "has invalid aggregate invocation type specification: " << keyword;
+}
+
+void printAggregationInvocation(
+    OpAsmPrinter &printer, CallOp op,
+    AggregationInvocationAttr aggregationInvocation) {
+  if (aggregationInvocation &&
+      aggregationInvocation.getValue() != AggregationInvocation::all) {
+    // The whitespace printed here compensates the trimming of whitespace in
```

Comment:
Can't one add it there ? E.g., use " " or some such? (just OOC, not blocker)

Reply:
I haven't found a way that makes me perfectly happy: if I add it there, I get *two* whitespaces in case this custom directive doesn't print anything. So now the declarative format suppresses a whitespace before this directive, then (1) if the directive prints something, it first prints the whitespace, then that thing, (2) otherwise it doesn't print anything, and in both cases, the declarative format adds another whitespace.